torch/csrc/jit/runtime/profiling_record.cpp

```diff
@@ -144,35 +144,42 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
   auto pno = pn->addOutput();
   pn->ty_(attr::profiled_type, TensorType::get());
   pno->setType(TensorType::get());
-  std::function<void(Stack&)> shape_profiler = [this, pno](Stack& stack) {
+  std::function<void(Stack&)> shape_profiler = [this, pn, pno](Stack& stack) {
     int64_t frame_id = 0;
     pop(stack, frame_id);
     IValue v;
     pop(stack, v);
+
     if (v.isTensor()) {
-      std::lock_guard<std::mutex> lock(this->mutex_);
-      auto& profiled_types = profiled_types_per_frame_[frame_id];
       auto& t = v.toTensor();
-      if (t.defined()) {
-        auto pttp = tensorTypeInCurrentExecutionContext(t);
+      auto new_tensor_type = tensorTypeInCurrentExecutionContext(t);
+
+      std::lock_guard<std::mutex> lock(this->mutex_);
+      if (profiling_count_ > 0) {
         GRAPH_DEBUG(
             "In run ",
             frame_id,
             " annotating %",
             pno->debugName(),
             " with ",
-            *pttp);
-        if (profiled_types.count(pno) == 0) {
-          profiled_types.insert({pno, pttp});
+            *new_tensor_type);
+
+        if (pn->hasRun()) {
```

Comment:
could we use `hasAttr(attr::profiled_type)` on `pn`?

Reply:
@Krovatkin there's a few places that expect `attr::profiled_type` to be set; I tried removing the initialization (i.e. `pn->ty_(attr::profiled_type, TensorType::get());`) but a bunch of tests fail, e.g.

```
ERROR: test_bias_as_arg (jit.test_autodiff_subgraph_slicing.TestAutodiffSubgraphSlicing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/dberard/pytorch/test/jit/test_autodiff_subgraph_slicing.py", line 135, in test_bias_as_arg
    check_against_reference(self, scripted, method1, lambda x: x, (x, weight, bias), check_types=False)
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/testing/_internal/common_jit.py", line 128, in check_against_reference
    grads_test = torch.autograd.grad(
  File "/home/dberard/local/miniconda3/envs/pytorch/lib/python3.9/site-packages/torch/autograd/__init__.py", line 275, in grad
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: required keyword attribute 'profiled_type' is undefined

----------------------------------------------------------------------
Ran 26 tests in 2.127s

FAILED (errors=1)
```

Reply:
@davidberard98 could you please elaborate why switching 
```
if (pn->hasRun() {
// ... merge types
} else {
 pn->ty_(attr::profiled_type, new_incoming_type);
}
```

to 

```
if (pn->hasAttribute(attr::profiled_type)) {
// ... merge types
} else {
 pn->ty_(attr::profiled_type, new_incoming_type);
 // next time we call `hasAttribute` we merge types
} 
```

Reply:
@Krovatkin 
Currently, we initialize `pn->ty_(attr::profiled_type, TensorType::get())` before the lambda gets called (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/profiling_record.cpp#L145). So if we simply replaced `hasRun()` with `hasAttribute()`, then `pn->hasAttribute(attr::profiled_type)` would always evaluate to true.

If we remove the initialization, then we get the `required keyword attribute 'profiled_type' is undefined` error that I mentioned.